pydantic_settings/sources.py

```diff
@@ -1679,6 +1692,48 @@ def _read_file(self, file_path: Path) -> dict[str, Any]:
             return yaml.safe_load(yaml_file)
 
 
+class AzureKeyVaultSettingsSource(EnvSettingsSource):
+    _url: str
+    _credential: TokenCredential  # type: ignore
+    _secret_client: SecretClient  # type: ignore
+
+    def __init__(
+        self,
+        settings_cls: type[BaseSettings],
+        url: str,
+        credential: TokenCredential,  # type: ignore
+        env_parse_none_str: str | None = None,
+        env_parse_enums: bool | None = None,
+    ) -> None:
+        import_azure_key_vault()
+        self._url = url
+        self._credential = credential
+        super().__init__(
+            settings_cls,
+            case_sensitive=True,
+            env_prefix='',
+            env_nested_delimiter='--',
+            env_ignore_empty=False,
+            env_parse_none_str=env_parse_none_str,
+            env_parse_enums=env_parse_enums,
+        )
+
+    def _load_env_vars(self) -> Mapping[str, str | None]:
+        self._secret_client = SecretClient(vault_url=self._url, credential=self._credential)  # type: ignore
+        secret_names: list[str] = [secret.name for secret in self._secret_client.list_properties_of_secrets()]
+        env_vars: dict[str, str | None] = {}
+
+        for secret_name in secret_names:
+            secret = self._secret_client.get_secret(secret_name)
+            secret_value = secret.value
+            env_vars[secret_name] = secret_value
```

Comment:
It loops over all the secrets and loads their values. I haven't tested it yet. do you know how long does it take if we have e.g 100 secrets?

Reply:
Less than 7 seconds.

It's the recommended option, and if you want a more advanced scenario or many more secrets, the normal is a) configure a refresh interval or b) use Azure App Configuration.

In other words, you instance your `BaseSettings` as a singleton when the backend is started.